openbadge-server/openbadge/views.py

```diff
@@ -372,48 +396,72 @@ def post_datafile(request, project_key):
             "details": "No data provided!",
             "chunks_written": 0,
             "chunks_received": 0
-        })
+        }, status_code=400)
 
     chunks = request.data.get("chunks")
-    # I don't like this but it works for now
+
     # sometimes we don't get json objects from the request object
     # (with tests, but I don't know if it happens anywhere else?)
+    # we need it in json instead of a list of strings so we can sort/operate on it
     if not isinstance(chunks, dict) and not isinstance(chunks, list):
         chunks = simplejson.loads(chunks)
 
+    # we sort the incoming data to be able to properly group the data
+    chunks = sorted(chunks, key=lambda chunk: chunk["data"]["timestamp"])
```

Comment:
We end up having 3 copies of the chunks. Could it create a memory problem in the server? Isn't that why we limited the file size to 15mb?

Three copies:
* chunks
* chunks = sorted(...) 
* grouped_chunks

We can get rid of one copy if we sort the chunk in place using list.sort()